src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -107,12 +107,15 @@ public static class ActivatorUtilities
 
                     if (isPreferred || bestLength < length)
                     {
+                        // When a preferred constructor is found, it is always selected however subsequent constructors
+                        // will be selected if they have more parameters.
```

Comment:
Do you think it is worth to do a breaking change and always be predictable and consistent regardless of the reflection order? 

I mean we can make the rule as either always select the preferred constructor or always select the constructor with longest parameters. 

I am not sure if this will be a bad breaking change but will be behavior will be more clear and predictable.

Reply:
> Do you think it is worth to do a breaking change and always be predictable and consistent regardless of the reflection order?

I don't know the full history here, but IMO yes a breaking change makes sense if we truly want to fix this issue.

There has only been [one minor change which was validation-related](https://github.com/dotnet/runtime/commit/f58cee20c39591835a01c2ffd5f2b3900b279c38) since the code [was moved from ASP.NET 6 years ago](https://github.com/dotnet/runtime/blob/d7704d4ec7a2f135d6a847da974d70213e5c09c9/src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs#L59-L79). One past issue is at https://github.com/dotnet/runtime/issues/42339 , which was closed as a dupe but shouldn't have been; I [added some content](https://github.com/dotnet/runtime/issues/42339#issuecomment-1977265887) to it that shows there is still a problem. Other related:
- Closed PR https://github.com/dotnet/runtime/pull/50688 which attempted to fix several issues.
- https://github.com/dotnet/runtime/pull/67493
- https://github.com/dotnet/runtime/issues/45119
- https://github.com/dotnet/runtime/issues/46132

Also, I believe the current implementation is intended to be high-performance, which is probably why it doesn't simply try to use logic such as:
```cs
if (!TryFindPreferredConstructor(...) &&  !TryFindMatchingConstructor(..)) { throw }
```
where TryFindMatchingConstructor does at least one alloc to order and match the parameters.

Reply:
I think the performance is important, if we go with any breaking change will be good not regress the performance.

Reply:
The second commit adds the breaking change; inro section updated.

Reply:
Also I discovered a perf improvement; I'll create a different PR for that.